src/label.rs

```diff
@@ -366,8 +371,42 @@ pub struct Label {
     pub polarity: Polarity,
     /// The path of the type being currently checked in the original type.
     pub path: ty_path::Path,
+    /// An environment mapping type variables to [`TypeVarData`]. Used by
+    /// polymorphic contracts to decide which actions to take when encountering a `forall`.
+    pub type_environment: HashMap<SealingKey, TypeVarData>,
 }
 
+/// Data about type variables that is needed for polymorphic contracts to decide which actions to take.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TypeVarData {
+    pub polarity: Polarity,
+}
+
+impl From<&TypeVarData> for Term {
+    fn from(value: &TypeVarData) -> Self {
+        Term::Record(RecordData {
+            fields: [(
+                Ident::new("polarity"),
+                Field::from(RichTerm::from(Term::from(value.polarity))),
+            )]
+            .into(),
+            attrs: Default::default(),
+            sealed_tail: None,
+        })
+    }
+}
+
+impl HasUnifType for TypeVarData {
```

Comment:
I find the name a tad misleading, I would expect a trait named `HasUnifType` to define a method `has_unif_type(&self) -> bool`. I don't know of a good name scheme for those kind of `* -> *` functions from a type to a type. `AssociatedUnifType`? Looks a bit long. `ReifyAsUnifType`? `ReifyAsNickelType`? I like reify, it sounds professional :sweat_smile:

Reply:
+1 for reify 👑 Another possibility would be `AsUnifType`. Really, the problem seems to be that `TypeVarData::unif_type()` should actually be an associated constant; and that the trait can't be named `UnifType` to conform to the usual naming format.

Reply:
Right, I would also have called it `UnifType` naturally.

Reply:
I've renamed it to `ReifyAsUnifType`. Yell if that's a bad idea 😅